Sources/Operators/Distinct.swift

```diff
@@ -0,0 +1,72 @@
+    /// De-duplicates _all_
+    /// published value events, as opposed to pairwise with `Publisher.removeDuplicates`.
+    ///
+    /// - note: It’s important to note that this operator stores all emitted values
+    ///         in an in-memory `Set`. So, use this operator with caution, when handling publishers
+    ///         that emit a large number of unique value events.
+    ///
+    /// - returns: A type-erased publisher that consumes duplicate values across all previous emissions from upstream.
+    func distinct() -> AnyPublisher<Output, Failure> {
+        scan((Set<Output>(), Output?.none)) { seenAndPrevious, incoming in
+            var seen = seenAndPrevious.0
+            let isUnique = seen.insert(incoming).inserted
+            return (seen, isUnique ? incoming : nil)
+        }
+        .compactMap { $0.1 }
+        .eraseToAnyPublisher()
+    }
+}
+
+public extension Publisher where Output: Equatable {
+    /// `Publisher.distinct` de-duplicates _all_
+    /// published `Hashable`-conforming value events, as opposed to pairwise with `Publisher.removeDuplicates`.
+    ///
+    /// - note: It’s important to note that this operator stores all emitted values
+    ///         in an in-memory `Array`. So, use this operator with caution, when handling publishers
+    ///         that emit a large number of unique value events.
+    ///
+    /// - returns: A type-erased publisher that consumes duplicate values across all previous emissions from upstream.
+    func distinct() -> AnyPublisher<Output, Failure> {
+        distinct(by: ==)
+    }
+}
+
+public extension Publisher {
+    /// `Publisher.distinct(by:)` de-duplicates _all_ published value events,
+    /// along the provided `by` predicate, as opposed to pairwise with `Publisher.removeDuplicates(by:)`.
+    ///
+    /// - parameter by: A predicate to use when determining uniqueness. `Publisher.distinct` will iterate
+    /// over all seen values applying each known unique value as the first argument to the predicate and the
+    /// incoming value event as the second, i.e. `by(see, next) -> Bool`. If this predicate is `true` for any
+    /// seen value, the next incoming value isn’t emitted downstream.
+    ///
+    /// - note: It’s important to note that this operator stores all emitted values
+    ///         in an in-memory `Array`. So, use this operator with caution, when handling publishers
+    ///         that emit a large number of unique value events (as per `by`).
+    ///
+    /// - returns: A type-erased publisher that consumes duplicate values across all previous emissions from upstream
+    ///            (signaled with `by`).
+    func distinct(by predicate: @escaping (Output, Output) -> Bool) -> AnyPublisher<Output, Failure> {
+        scan(([Output](), Output?.none)) { seenAndPrevious, incoming in
+            var seen = seenAndPrevious.0
+
+            guard !seen.contains(where: { predicate($0, incoming) }) else { return (seen, nil) }
```

Comment:
nit: Personally, I'm not a fan of guarding against a negation, I find it a bit harder to reason. 
I think `if` instead  would make this a bit more simple 😊

Reply:
Agreed, updating this in the next commit. Thanks. 🙏🏽